src/platform/posix/tcp_ip_channel.c

```diff
@@ -142,25 +140,25 @@ TaggedMessage *TcpIpChannel_receive(NetworkChannel *untyped_self) {
     socket = self->fd;
   }
 
-  if (self->blocking) {
-    ioctl(socket, FIONREAD, &bytes_available);
-    bytes_available = MIN(8, bytes_available);
-  } else {
-    // peek into file descriptor to figure how many bytes are available.
-    ioctl(socket, FIONREAD, &bytes_available);
-  }
-
-  if (bytes_available == 0) {
+  // configure select to check if data is available (non blocking)
+  fd_set read_fds;
+  struct timeval timeout;
+  FD_ZERO(&read_fds);
+  FD_SET(socket, &read_fds);
+  timeout.tv_sec = 0;
+  timeout.tv_usec = 0;
+
+  // check if data is available
+  int select_result = select(socket + 1, &read_fds, NULL, NULL, &timeout);
+  if (select_result <= 0) {
```

Comment:
Do we actually need to check how much data is there before we read? Since this is running in a separate thread, we want it to block on `recv` and return with as much data is available. What is gained from checking how much data is there before the read? 

(I know that this was there before your edits, and perhaps @tanneberger can answer better?)